core/src/main/java/org/elasticsearch/common/util/concurrent/QueueResizingEsThreadPoolExecutor.java

```diff
@@ -0,0 +1,161 @@
+    private final int tasksPerWindow;
+    private final int minQueueSize;
+    private final int maxQueueSize;
+
+    private final AtomicLong totalTaskNanos = new AtomicLong(0);
+    private final AtomicInteger taskCount = new AtomicInteger(0);
+
+    private long startNs;
+
+    QueueResizingEsThreadPoolExecutor(String name, int corePoolSize, int maximumPoolSize,
+                                      long keepAliveTime, TimeUnit unit,
+                                      ResizableSizeBlockingQueue<Runnable> workQueue,
+                                      int minQueueSize, int maxQueueSize,
+                                      Function<Runnable, Runnable> runnableWrapper,
+                                      final int tasksPerWindow,
+                                      ThreadFactory threadFactory,
+                                      XRejectedExecutionHandler handler,
+                                      ThreadContext contextHolder) {
+        super(name, corePoolSize, maximumPoolSize, keepAliveTime, unit,
+                workQueue, threadFactory, handler, contextHolder);
+        this.runnableWrapper = runnableWrapper;
+        this.workQueue = workQueue;
+        this.tasksPerWindow = tasksPerWindow;
+        this.startNs = System.nanoTime();
+        this.minQueueSize = minQueueSize;
+        this.maxQueueSize = maxQueueSize;
+    }
+
+    @Override
+    protected void doExecute(final Runnable command) {
+        // we are submitting a task, it has not yet started running (because super.excute() has not
+        // been called), but it could be immediately run, or run at a later time. We need the time
+        // this task entered the queue, which we get by creating a TimedRunnable, which starts the
+        // clock as soon as it is created.
+        super.doExecute(this.runnableWrapper.apply(command));
+    }
+
+    static int calculateL(final long totalRuntime, final int taskCount,
+                          final long totalTaskNanos, final String name,
+                          final Logger logger) {
+        assert totalRuntime > 0 : "cannot calculate for instantaneous tasks";
+        assert taskCount > 0 : "cannot calculate for no tasks";
+        // There is no set execution time, instead we adjust the time window based on the
+        // number of completed tasks, so there is no backwards thread required to update the
+        // queue size at a regular interval. This means we need to calculate our λ by the
+        // total runtime, rather than a fixed interval.
+
+        // λ = tasks per time period - total tasks divided by measurement time
+        final double lambda = (double) taskCount / totalRuntime;
+
+        // W = avg task execution time - total task execution divided by total task count
+        final double w = totalTaskNanos / taskCount;
+
+        // L = λ * W
+        final int l = (int) (lambda * w);
+        logger.debug("estimated optimal queue size for [{}] threadpool: " +
```

Comment:
do we need this logging statement of can we move it to the consumer of this method? I'd love to not have any logging in here.